pkg/volume/csi/csi_driver_reg.go

```diff
@@ -0,0 +1,288 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package csi
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/golang/glog"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/util/retry"
+)
+
+const (
+	// Name of node annotation that contains JSON map of driver names to node
+	// names
+	annotationKey = "csi.volume.kubernetes.io/nodeid"
+)
+
+type csiDriver struct {
+	driverName     string
+	driverEndpoint string
+}
+
+// csiDrivers map keep track of all registered CSI drivers on the node and their
+// corresponding sockets
+var csiDrivers sync.Map
+
+// labelManagement is struct of channels used for communication between the driver registration
+// code and the go routine responsible for managing the node's labels
+type labelManagement struct {
+	in  chan csiDriver
+	out chan bool
+}
+
+var lm labelManagement
+
+// Initalizing csiDrivers map and label management channels
+func init() {
+	csiDrivers = sync.Map{}
+	lm = labelManagement{in: make(chan csiDriver), out: make(chan bool)}
+}
+
+// nodeLabelManager waits for labeling requests initiated by the driver's registration
+// process.
+// TODO (sbezverk) Right now the driver registration only registers drivers, not clear
+// how to de-register the driver when CSI driver is terminated.
+// TODO (sbezverk) Check for pluginwatcher behavior on kubelet restart, CSI would require
+// replaying registrations for all existing socket. (Possible issue is when kubelet is up and tryig
+// to replay registrations but corresponding driver-registrars are not up yet, would it keep retrying?)
+func nodeLabelManager(nodeName types.NodeName, k8s kubernetes.Interface) {
```

Comment:
Convert this function to method for type labelManager i.e.
```go
func (m *labelManager) Run(...){

}
```

Reply:
+1 create a interface for labelManager. Create a `newLabelManager` method that initializes the `labelManager struct`. And turn this function in to `Run` as suggested by Vlad. All reads and writes to the `labelManager struct` should go through this methods in this new class instead of directly on the struct. 

A good example to follow: https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/nestedpendingoperations/nestedpendingoperations.go

Reply:
`newLabelManager` can accept `NodeName` and `KubeClient` as parameters and store them in its internal struct.